File: packages/guard/src/core/decision.ts
import type { Decision, Finding, Severity } from "../types/index.js";

const decisionOrder: Record<Decision, number> = {
  allow: 0,
  require_approval: 1,
  block: 2
};

function maxDecision(a: Decision, b: Decision): Decision {
  return decisionOrder[a] >= decisionOrder[b] ? a : b;
}

export function severityToDecision(severity: Severity): Decision {
  if (severity === "high") return "block";
  if (severity === "med") return "require_approval";
  return "allow";
}

export function decisionFromFindings(findings: Finding[], fallback: Decision = "allow"): Decision {
  let decision = fallback;
  for (const finding of findings) {
    decision = maxDecision(decision, finding.action);
  }
  return decision;
}

export function mergeDecisionWithAnomaly(policyDecision: Decision, anomalyProposal: Decision): {
  decision: Decision;
  escalatedByAnomaly: boolean;
} {
  if (policyDecision === "block") {
    return { decision: "block", escalatedByAnomaly: false };
  }

  if (anomalyProposal === "allow") {
    return { decision: policyDecision, escalatedByAnomaly: false };
  }

  if (anomalyProposal === "require_approval") {
    if (policyDecision === "allow") {
      return { decision: "require_approval", escalatedByAnomaly: true };
    }
    return { decision: policyDecision, escalatedByAnomaly: false };
  }

  if (policyDecision === "allow") {
    return { decision: "require_approval", escalatedByAnomaly: true };
  }

  if (policyDecision === "require_approval") {
    return { decision: "block", escalatedByAnomaly: true };
  }

  return { decision: policyDecision, escalatedByAnomaly: false };
}

File: packages/python/src/twire_guard/decision.py
from __future__ import annotations

from typing import Any

DECISION_ORDER = {"allow": 0, "require_approval": 1, "block": 2}


def _max_decision(a: str, b: str) -> str:
    return a if DECISION_ORDER[a] >= DECISION_ORDER[b] else b


def severity_to_decision(severity: str) -> str:
    if severity == "high":
        return "block"
    if severity == "med":
        return "require_approval"
    return "allow"


def decision_from_findings(findings: list[dict[str, Any]], fallback: str = "allow") -> str:
    decision = fallback
    for finding in findings:
        decision = _max_decision(decision, str(finding["action"]))
    return decision


def merge_decision_with_anomaly(policy_decision: str, anomaly_proposal: str) -> dict[str, Any]:
    if policy_decision == "block":
        return {"decision": "block", "escalated_by_anomaly": False}

    if anomaly_proposal == "allow":
        return {"decision": policy_decision, "escalated_by_anomaly": False}

    if anomaly_proposal == "require_approval":
        if policy_decision == "allow":
            return {"decision": "require_approval", "escalated_by_anomaly": True}
        return {"decision": policy_decision, "escalated_by_anomaly": False}

    if policy_decision == "allow":
        return {"decision": "require_approval", "escalated_by_anomaly": True}

    if policy_decision == "require_approval":
        return {"decision": "block", "escalated_by_anomaly": True}

    return {"decision": policy_decision, "escalated_by_anomaly": False}
